### PR TITLE
regru: client certificate support

### PR DESCRIPTION
**Pull request by boiler**
Added support for "reg.ru" DNS provider to additional authentication by SSL certificates pair

### CODE REVIEW

File: providers/dns/regru/regru.go [hunk 1]
```diff
@@ -87,6 +94,18 @@ func NewDNSProviderConfig(config *Config) (*DNSProvider, error) {
 		client.HTTPClient = config.HTTPClient
 	}
 
+	if config.TlsCert != "" || config.TlsKey != "" {
```
Comment (ldez) [inline providers/dns/regru/regru.go:97]:
```suggestion
	if config.TlsCert != "" && config.TlsKey != "" {
```

File: providers/dns/regru/regru.go [hunk 2]
```diff
@@ -87,6 +94,18 @@ func NewDNSProviderConfig(config *Config) (*DNSProvider, error) {
...
+		tlsCert, err := tls.X509KeyPair([]byte(config.TlsCert), []byte(config.TlsKey))
+		if err != nil {
+			return nil, errors.New("regru: " + err.Error())
```
Comment (ldez) [inline providers/dns/regru/regru.go:100]:
```suggestion
			return nil, fmt.Errorf("regru: %w", err)
```

**Review by ldez — CHANGES_REQUESTED** [review]

File: providers/dns/regru/regru.go [hunk 3]
```diff
@@ -17,8 +18,10 @@ import (
 const (
 	envNamespace = "REGRU_"
 
-	EnvUsername = envNamespace + "USERNAME"
-	EnvPassword = envNamespace + "PASSWORD"
+	EnvUsername    = envNamespace + "USERNAME"
+	EnvPassword    = envNamespace + "PASSWORD"
+	EnvTlsCertPath = envNamespace + "SSL_CERT"
+	EnvTlsKeyPath  = envNamespace + "SSL_KEY"
```
Comment (ldez) [inline providers/dns/regru/regru.go:24]:
```suggestion
	EnvTLSCertPath = envNamespace + "TLS_CERT"
	EnvTLSKeyPath  = envNamespace + "TLS_KEY"
```

File: providers/dns/regru/regru.go [hunk 4]
```diff
@@ -30,6 +33,8 @@ const (
 type Config struct {
 	Username string
 	Password string
+	TlsCert  string
+	TlsKey   string
```
Comment (ldez) [inline providers/dns/regru/regru.go:37]:
```suggestion
	TLSCert  string
	TLSKey   string
```

File: providers/dns/regru/regru.go [hunk 5]
```diff
@@ -40,6 +45,8 @@ type Config struct {
 // NewDefaultConfig returns a default configuration for the DNSProvider.
 func NewDefaultConfig() *Config {
 	return &Config{
+		TlsCert:            env.GetOrDefaultString(EnvTlsCertPath, ""),
+		TlsKey:             env.GetOrDefaultString(EnvTlsKeyPath, ""),
```
Comment (ldez) [inline providers/dns/regru/regru.go:49]:
```suggestion
		TLSCert:            env.GetOrDefaultString(EnvTLSCertPath, ""),
		TLSKey:             env.GetOrDefaultString(EnvTLSKeyPath, ""),
```

File: providers/dns/regru/regru.go [hunk 6]
```diff
@@ -87,6 +94,18 @@ func NewDNSProviderConfig(config *Config) (*DNSProvider, error) {
 		client.HTTPClient = config.HTTPClient
 	}
 
+	if config.TlsCert != "" || config.TlsKey != "" {
+		tlsCert, err := tls.X509KeyPair([]byte(config.TlsCert), []byte(config.TlsKey))
```
Comment (ldez) [inline providers/dns/regru/regru.go:98]:
```suggestion
		tlsCert, err := tls.X509KeyPair([]byte(config.TLSCert), []byte(config.TLSKey))
```

File: providers/dns/regru/regru.toml [hunk 7]
```diff
@@ -15,6 +15,8 @@ lego --email you@example.com --dns regru --domains my.example.org run
     REGRU_USERNAME = "API username"
     REGRU_PASSWORD = "API password"
   [Configuration.Additional]
+    REGRU_SSL_CERT = "SSL authentication certificate"
+    REGRU_SSL_KEY = "SSL authentication private key"
```
Comment (ldez) [inline providers/dns/regru/regru.toml:19]:
```suggestion
    REGRU_TLS_CERT = "authentication certificate"
    REGRU_TLS_KEY = "authentication private key"
```

**Review by ldez — CHANGES_REQUESTED** [review]

File: providers/dns/regru/regru.go (hunk 1 above)
Comment (boiler) [reply providers/dns/regru/regru.go:97]: If user makes a mistake and defines only one variable he will see an error
With your suggestion he will see nothing

Comment (ldez) [reply providers/dns/regru/regru.go]: I think that an explicit error is better:
```go
	
	if config.TLSCert != "" && config.TLSKey == "" || config.TLSCert == "" && config.TLSKey != "" {
		return nil, errors.New("regru: certificate or key is missing")
	}

	if config.TLSCert != "" && config.TLSKey != "" {
	// ...
```

Comment (boiler) [reply providers/dns/regru/regru.go]: What about? :

```go
if config.TLSCert != "" || config.TLSKey != "" {
        if config.TLSCert == "" {
		return nil, errors.New("regru: TLS certificate is missing")
	}
        if config.TLSKey == "" {
		return nil, errors.New("regru: TLS key is missing")
	}
```

Comment (ldez) [reply providers/dns/regru/regru.go]: yes, it's possible, it will increase the cyclomatic complexity but it will work.

**Review by ldez — APPROVED** [review]
LGTM

**Pull request merged**